### PR TITLE
Merge testsuite sub-makefiles in the root Makefile and harden build system

### LINKED ISSUE #12587: Test failure when glibc has frame pointers
I attempted a test build of OCaml 5.1.0 for the Fedora Linux distribution, but all of the frame-pointers tests failed.  The first failure looks like this:
```
Running tests from 'tests/frame-pointers' ...
 ... testing 'c_call.ml' with 1 (frame_pointers) => passed
 ... testing 'c_call.ml' with 1.1 (native) sh: line 1: : command not found
=> failed (program output /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/c_call.opt.output differs from reference /builddir/build/BUILD/ocaml-5.1.0/testsuite/tests/frame-pointers/c_call.reference: 
)
> Specified modules: c_call.ml
> Source modules: c_call.ml
> Running test frame_pointers with 1 actions
> 
> Running action 1/1 (frame_pointers)
> Action 1/1 (frame_pointers) => passed (frame-pointers available)
> Running test native with 8 actions
> 
> Running action 1/8 (setup-ocamlopt.byte-build-env)
> Action 1/8 (setup-ocamlopt.byte-build-env) => passed
> 
> Running action 2/8 (ocamlopt.byte)
> Compiling program /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/c_call.opt from modules  fp_backtrace.c c_call_.c c_call.ml
> Commandline: /builddir/build/BUILD/ocaml-5.1.0/runtime/ocamlrun /builddir/build/BUILD/ocaml-5.1.0/ocamlopt -I /builddir/build/BUILD/ocaml-5.1.0/runtime -ccopt -I/builddir/build/BUILD/ocaml-5.1.0/runtime -nostdlib -I /builddir/build/BUILD/ocaml-5.1.0/stdlib -o /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/c_call.opt fp_backtrace.c c_call_.c c_call.ml
>   Redirecting stdout to /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/ocamlopt.byte.output 
>   Redirecting stderr to /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/ocamlopt.byte.output 
> Action 2/8 (ocamlopt.byte) => passed
> 
> Running action 3/8 (check-ocamlopt.byte-output)
> Comparing compiler output /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/ocamlopt.byte.output to reference /builddir/build/BUILD/ocaml-5.1.0/testsuite/tests/frame-pointers/c_call.compilers.reference
> Action 3/8 (check-ocamlopt.byte-output) => passed
> 
> Running action 4/8 (run)
> Entering run_hook for hook /builddir/build/BUILD/ocaml-5.1.0/testsuite/tests/frame-pointers/c_call.run
> Hook should write its response to /tmp/ocamltest-1d976d.response
> Action 4/8 (run) => passed
> 
> Running action 5/8 (check-program-output)
> Comparing program output /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/c_call.opt.output to reference /builddir/build/BUILD/ocaml-5.1.0/testsuite/tests/frame-pointers/c_call.reference
> Action 5/8 (check-program-output) => failed (program output /builddir/build/BUILD/ocaml-5.1.0/testsuite/_ocamltest/tests/frame-pointers/c_call/ocamlopt.byte/c_call.opt.output differs from reference /builddir/build/BUILD/ocaml-5.1.0/testsuite/tests/frame-pointers/c_call.reference: 
> 
> )
```

That `sh: line 1: : command not found` text had me scrambling all over the makefiles, looking for some command being invoked that is not present in the buildroot.  That turned out to be a red herring.  The real problem is that the expected and actual outputs differ, and the `filecompare.ml` code reported the difference.  When that happens the difference is reported as failure to find a command, which is misleading.  But that's not what this issue is about.

Fedora builds all amd64 packages with frame pointers.  The output of the `c_call` test looks like this:
```
fp_backtrace_many_args
caml_c_call_stack_args
camlC_call.f
camlC_call.entry
caml_program
caml_start_program
caml_main/caml_startup
main
_start
caml_c_call
camlC_call.f
camlC_call.entry
caml_program
caml_start_program
caml_main/caml_startup
main
_start
camlC_call.f
camlC_call.entry
caml_program
caml_start_program
caml_main/caml_startup
main
_start
```

This differs from the expected output because each stack trace has `_start` after `main`, due to glibc having been built with frame pointers.  The frame-pointers tests should not fail because of the presence of `_start`.

### PR DESCRIPTION
**Pull request by shindere**
This PR merges `testsuite/{lib,tools}/Makefile` into the root Makefile and does two additional perhaps more interesting things:

 1. If ocamltest is enabled, all the support libraries and tools required to run certain tests are built, too. This means that any test can then be run by ocamltest directly without having to necessarily go through make to ensure that the required libraries and tools are built. This fulfulls a wish expressed some time ago by @Octachron and I am pleased to offer this modest gift to him. @gasche may also like this, and the coming item, too.

 2. The PR makes it impossible to build ocamltest if it has not been enabled at configure time and gives an explicit error message explaining what is going on. This is an attempt to address #12587, more precisely a problem it revealed, see e.g. https://github.com/ocaml/ocaml/issues/12587#issuecomment-175952364 since if ocamltest is disabled at configuration time, this PR makes is then completely impossible to build it, even manually.

### CODE REVIEW

File: testsuite/lib/testing.ml [hunk 1]
```diff
@@ -86,11 +86,13 @@ let test_raises_this_failure s f x =
...
 (* Applies f to x and checks that the evaluation indeed
    raises the exception Failure. *)
+(*
 let test_raises_some_failure f x =
   test_raises_exc_p (function Failure _ -> true | _ -> false) f x;;
+*)
 
 let failure_test f x s = test_raises_this_failure s f x;;
-let any_failure_test = test_raises_some_failure;;
+(* let any_failure_test = test_raises_some_failure;; *)
```
Comment (dra27) [inline testsuite/lib/testing.ml]: Instead of commenting this out, why not expose it in the .mli?

File: testsuite/lib/lib.mli [hunk 2]
```diff
@@ -0,0 +1,57 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                      Sebastien Hinderer, Tarides                       *)
```
Comment (dra27) [inline testsuite/lib/lib.mli]: FWIW I'd use the copyright header of lib.ml, given that the file is largely identical

File: Makefile [hunk 3]
```diff
@@ -1874,6 +1903,11 @@ partialclean::
 	rm -rf $(ocamltest_GENERATED_FILES)
 	rm -f ocamltest/ocamltest.html
 	rm -f $(addprefix testsuite/lib/*.,cm* o obj a lib)
+	rm -f $(addprefix testsuite/tools/*.,cm* o obj a lib)
+	rm -f testsuite/tools/codegen testsuite/tools/codegen.exe
+	rm -f testsuite/tools/expect testsuite/tools/expect.exe
+	rm -f testsuite/tools/lexcmm.ml # Documentation
```
Comment (dra27) [inline Makefile]: The "# Documentation" looks like an editing artefact?

File: Makefile [hunk 4]
```diff
@@ -1832,7 +1833,7 @@ $(ocamltest_DEP_FILES): $(DEPDIR)/ocamltest/%.$(D): ocamltest/%.c
 ocamltest/%: CAMLC = $(BEST_OCAMLC) $(STDLIBFLAGS)
 
 ocamltest: ocamltest/ocamltest$(EXE) \
-  testsuite/lib/lib.cmo testsuite/lib/testing.cma
+  testsuite/lib/lib.cmo testsuite/lib/testing.cma testsuite/tools/expect$(EXE)
```
Comment (dra27) [inline Makefile:1812]: I don't think this needs to be here - isn't testsuite/tools/expect built as part of BYTECODE_OPTIONAL_TOOLS?

File: Makefile [hunk 5]
```diff
@@ -1842,11 +1843,39 @@ testing_LIBRARIES =
...
+	$(V_ASM)$(ASPP) $(OC_ASPPFLAGS) -o $@ $< || $(ASPP_ERROR)
+endif
+
 ocamltest/ocamltest$(EXE): OC_BYTECODE_LINKFLAGS += -custom
 
 ocamltest/ocamltest$(EXE): ocamlc ocamlyacc ocamllex
 
-ocamltest.opt: ocamltest/ocamltest.opt$(EXE) testsuite/lib/testing.cmxa
+ocamltest.opt: ocamltest/ocamltest.opt$(EXE) \
+  testsuite/lib/testing.cmxa $(asmgen_OBJECT) testsuite/tools/codegen$(EXE)
```
Comment (dra27) [inline Makefile:1854]: Similarly, isn't this part of the optional tools variable?

File: configure.ac [hunk 6]
```diff
@@ -2314,7 +2319,8 @@ AS_CASE([$enable_ocamltest,OCAML__DEVELOPMENT_VERSION],
     ocamltest_target=ocamltest
     ocamltest_opt_target=ocamltest.opt
     ocamltest='ocamltest'
-    ],
+    testsuite_tools='testsuite/tools/codegen testsuite/tools/expect'
```
Comment (dra27) [inline configure.ac:2322]: As a further optimisation (I think) codegen is only needed if the native compiler is being built

**Review by dra27 — COMMENTED** [review]
A few suggestions here and there, with a couple of other thoughts:
- Given the integration into the main Makefile, could we just move expect, codegen, testing and "lib" (which could be given a more descriptive name) to `tools/` (in the same way that stripdebug, cmpbyt and other parts of the build system also live in tools)?
- I fully support hardening the correct running of the testsuite, but I'm not overly keen on this approach. It is relatively easy to end up working on a tree where ocamltest wasn't built by default, and it's very pedantic and time-consuming to force the user to reconfigure the tree. I would prefer it to be that the test for, say, `diff` and everything is _always_ done, but is only an error if `ocamltest` is explicitly _requested_ and, consequently, that attempting to build `ocamltest` is only an error if it _impossible_ to build.

**Comment by shindere** [comment]
David Allsopp (2023/11/11 06:07 -0800):
>      * Given the integration into the main Makefile, could we just move
>        expect, codegen, testing and "lib" (which could be given a more
>        descriptive name) to tools/ (in the same way that stripdebug,
>        cmpbyt and other parts of the build system also live in tools)?

This had actually already been discussed in the past, although I can't
remember whether the discussion took place online or off-line. At that
time, the conclusion was that we wanted to keep the tools and libs
related to tests separated from the rest but that decision can of course
always be revisited.

>      * I fully support hardening the correct running of the testsuite, but
>        I'm not overly keen on this approach. It is relatively to end up
>        working on a tree where ocamltest wasn't built by default, and it's
>        very pedantic and time-consuming to force the user to reconfigure
>        the tree. I would prefer it to be that the test for, say, diff and
>        everything is always done, but is only an error if ocamltest is
>        explicitly requested and, consequently, that attempting to build
>        ocamltest is only an error if it impossible to build.

David and I discussed this point intensively offline. My understanding
of the conclusion is that, for the time being, the fact that there is an
error hinting the user on what to do is already an improvement over te
previous situation. Ideally, though, in the case of ocamltest (and only
in that case), it seems desirable to make it possible to build it
although it has been disabled at configure time (provided it's actually
possible to do so), so that workflows like `./configure && make -- make
test` work in any case. That means that if ocamltest is enabled then it
will be build during the main build, but if it's not, then it will be
build as part of `make test`. At the moment this is non-trivial to
implement so we think it's best to wait until `testsuite/Makefile` gets
merged into the root Makefile too to implement this feature. I'll let
@dra27 comment on whether this summary of our discussion is accurate.

>    In testsuite/lib/testing.ml:
> >
>  let failure_test f x s = test_raises_this_failure s f x;;
> -let any_failure_test = test_raises_some_failure;;
> +(* let any_failure_test = test_raises_some_failure;; *)
> 
>    Instead of commenting this out, why not expose it in the .mli?

I did so to mark that this bit of code is currently unused.

Given this explanation, what do you prefer? Shall I leave it as it is or
expose it?

>    In testsuite/lib/lib.mli:
> > @@ -0,0 +1,57 @@
> +(**************************************************************************)
> +(*                                                                        *)
> +(*                                 OCaml                                  *)
> +(*                                                                        *)
> +(*                      Sebastien Hinderer, Tarides                       *)
> 
>    FWIW I'd use the copyright header of lib.ml, given that the file is
>    largely identical

OK, Copyright adjusted

>    In Makefile:
> > @@ -1874,6 +1903,11 @@ partialclean::
>         rm -rf $(ocamltest_GENERATED_FILES)
>         rm -f ocamltest/ocamltest.html
>         rm -f $(addprefix testsuite/lib/*.,cm* o obj a lib)
> +       rm -f $(addprefix testsuite/tools/*.,cm* o obj a lib)
> +       rm -f testsuite/tools/codegen testsuite/tools/codegen.exe
> +       rm -f testsuite/tools/expect testsuite/tools/expect.exe
> +       rm -f testsuite/tools/lexcmm.ml # Documentation
> 
>    The "# Documentation" looks like an editing artefact?

Indeed! Might be the one a few lines down which get copied I don't know
how. Fixed.

>    In Makefile:
> > @@ -1832,7 +1833,7 @@ $(ocamltest_DEP_FILES): $(DEPDIR)/ocamltest/%.$(D): ocam
> ltest/%.c
>  ocamltest/%: CAMLC = $(BEST_OCAMLC) $(STDLIBFLAGS)
> 
>  ocamltest: ocamltest/ocamltest$(EXE) \
> -  testsuite/lib/lib.cmo testsuite/lib/testing.cma
> +  testsuite/lib/lib.cmo testsuite/lib/testing.cma testsuite/tools/expect$(EXE)
> 
>    I don't think this needs to be here - isn't testsuite/tools/expect
>    built as part of BYTECODE_OPTIONAL_TOOLS?

You are right that there is some redundancy here that needs to be
addressed, many thanks for having spotted it. I am not sure yet how to
best address it though, i.e. whether I should remove the dependency, or
rather adjust `configure.ac`.

One interesting thing to notice about `codegen` is that, at least for
the time being,
it's a bytecode tool but that is used only for native tests. Its
situation thus looks a bit odd to me. Wouldn't it make things a bit
clearer if we would build it as a native tool only, and then only when
the native compiler is enabled?

Then of course the question of how to build it remains. Ot me, having
the tools as dependencies of ocamltest looks slightly more meaningful
than adding them to generic list of tools. What's your take?

>    In Makefile:
> >  ocamltest/ocamltest$(EXE): OC_BYTECODE_LINKFLAGS += -custom
> 
>  ocamltest/ocamltest$(EXE): ocamlc ocamlyacc ocamllex
> 
> -ocamltest.opt: ocamltest/ocamltest.opt$(EXE) testsuite/lib/testing.cmxa
> +ocamltest.opt: ocamltest/ocamltest.opt$(EXE) \
> +  testsuite/lib/testing.cmxa $(asmgen_OBJECT) testsuite/tools/codegen$(EXE)
> 
>    Similarly, isn't this part of the optional tools variable?

I see the response as a matter of taste, as I tried to explain above,
but let's see.

>    In [6]configure.ac:
> > @@ -2314,7 +2319,8 @@ AS_CASE([$enable_ocamltest,OCAML__DEVELOPMENT_VERSION],
>      ocamltest_target=ocamltest
>      ocamltest_opt_target=ocamltest.opt
>      ocamltest='ocamltest'
> -    ],
> +    testsuite_tools='testsuite/tools/codegen testsuite/tools/expect'
> 
>    As a further optimisation (I think) codegen is only needed if the
>    native compiler is being built

Yes, I think it's what I tried to do by adding
`testsuite/tools/codegen$(EXE)` as a dependency of `ocamltest.opt` in
the root Makefile but perhaps that's not clean enough?

**Comment by shindere** [comment]
Giving it a second thought, I believe that the code in this PR is actually correct as it is.

More specifically, regarding the discussions on where we should list the
libraries and tools to be built, the lists in `configure.ac` and the
dependencies in `Makefile` accomplish two different purposes. The fact that
the libraries and tools are listed in `configre.ac` is useful so that the
macros for building programs and libraries get called. But these macros
being called is not enough per se for the corresponding libraries and tools
to be actually built. The build will become effective only if the libraries
and tools appear as prerequisites of other rules, which is what is done in
the Makefile.

**Comment by shindere** [comment]
I just pushed an update that takes into account @dra27's remark about
the Testing library module.

With this, I hope I have addressed all the review comments done so far.

**Review by dra27 — APPROVED** [review]
LGTM!

**Comment by shindere** [comment]
Thanks! And thanks for the careful and supportive review, as usual!

**Pull request merged**